First steps towards x86 compatible crossgen2

davidwrighton opened the pull request:
- Fix bug in crossgen2 around compilation of type forwarder only assemblies
- Use 64bit x86 crossgen2 when running crossgen2 as part of test build
- Disable inlined pinvoke generation on X86 until failures are understood
- Autolayout alignment should align to 4 byte offsets on X86 platforms
- Convert ArgIterator to use CoreCLR ABI instead of ProjectN ABI
- Align instance fields of classes to the start of the fields of the object, not the start of the object
- Fix hardware intrinsic support for X86
- MDArray address method uses custom calling convention on X86
- X86 JIT expects VM to normalize some single element structures to int, and inptr types
- Enable crossgen2smoke test for x86

With this change the crossgen2smoke test passes when compiled with crossgen2, as do simpler tests which use a crossgen2 compiled variant of the framework libraries. crossgen2smoke + libraries compiled with crossgen2 does not pass due to issues in resource handling.

davidwrighton committed:
Fix compilation of type forwarder assemblies
- Also adjust build of crossgen2 binaries in core root preparation to work for x86

davidwrighton committed:
Fix argument layout handling

davidwrighton committed:
Disable inline pinvokes until x86 handling is understood

davidwrighton committed:
Disable excess alignment for autolayout on X86

davidwrighton committed:
Refactoring oops around GCRefMapCreation

davidwrighton committed:
Align fields on X86 assuming that the start of the fields on an instance object is always aligned

davidwrighton committed:
The multidimensional array address helper on X86 has a special calling convention which takes a param arg when working with shared code

davidwrighton committed:
Enable crossgen2smoke test for X86

davidwrighton committed:
Fix hardware intrinsic support for X86 targeted builds

davidwrighton commented:
@dotnet/crossgen-contrib I took a break from working on infrastructure, and put together a few fixes around support for X86 targeted compilation.

trylek commented on src/tests/readytorun/crossgen2/Program.cs:
```diff
@@ -1537,7 +1542,9 @@ private unsafe static bool ExplicitlySizedStructTest()
                 return false;
         }
 
+        if (sizeof(IntPtr) == 8) // This test isn't right for 32 bit platforms
```
Hmm, I'm somewhat confused here - you're disabling the test on x86 yet you changed the <code>Size</code> of the <code>StructLayout</code> attribute on <code>ExplicitlySizedStructAuto</code> - does that mean that the previous value was wrong or is the change from 0x14 to 0x18 supposed to be more stressful? I don't think it's a big deal, it just comes as a surprise why you need to modify a test you're disabling for the "new x86 scenario" you're adding.

davidwrighton replied:
Well, the test is just wrong. It turns out it doesn't pass when run without crossgen2, with exactly the same fault, but by putting the disable logic as IL in the test, we actually get testing that the behavior of the x86 type loader at runtime matches up with the behavior at crossgen2 time due to the presence of the --verify-type-and-field-layout switch that we use when compiling crossgen2 tests.

I agree though, I was quite disappointed to find that I needed to change the test.

As far as how much was passing, as I mentioned to you a couple days ago, I was looking for something different than endless infrastructure stuff, so I took a look at this, and its come pretty clean. At this point I believe our type layout is very nearly correct, as is our argument layout handling, which leaves whatever interesting details are special about x86 in the jit interface. I have some suspicions that then handling of single element struct normalization isn't quite right, but its not all that far off from right.

trylek reviewed (approved):
I don't pretend I'm fluent in all the x86 calling convention details but overall the change looks awesome to me, thanks David, I had no idea x64->x86 JIT support is functional enough to let you pass most of the smoke test, that's great news!

davidwrighton committed:
Fix il verification build break

MichalStrehovsky commented on src/coreclr/src/tools/Common/Compiler/CompilerTypeSystemContext.cs:
```diff
@@ -330,6 +331,68 @@ private PdbSymbolReader OpenAssociatedSymbolFile(string peFilePath, PEReader peR
+        private HashSet<TypeDesc> _nonNormalizedTypes = new HashSet<TypeDesc>();
+        private Dictionary<TypeDesc, TypeFlags> _normalizedTypeCategory = new Dictionary<TypeDesc, TypeFlags>();
+
+        public TypeFlags NormalizedCategoryFor4ByteStructOnX86(TypeDesc type)
```
There's a discussion about adding x86 support to Native AOT - one of the the things we would want to do is make RyuJIT changes so that we don't have to compensate for 20 year old legacy quirks in x86 RyuJIT. Do you know if this is one of such quirks? If so, I would prefer we put it in a R2R-specific file.

MichalStrehovsky commented on src/coreclr/src/tools/Common/Compiler/CompilerTypeSystemContext.cs:
```diff
@@ -330,6 +331,68 @@ private PdbSymbolReader OpenAssociatedSymbolFile(string peFilePath, PEReader peR
+                    throw new NotSupportedException();
+                }
+
+                if ((type.Category != TypeFlags.ValueType) || (type.GetElementSize().AsInt != 4))
```
This check is cheap. Maybe we only need to cache this for 4-byte valuetypes and we can get by without locking/caching otherwise? Might also be worth to frontload the I4/U4/I/U check so that we can do all the common things without taking the lock and doing hashtable lookups.

davidwrighton replied:
I moved this check up. The rest might/might not be valuable. I'll wait for a profile for the rest.

MichalStrehovsky commented on src/coreclr/src/tools/Common/Compiler/CompilerTypeSystemContext.cs:
```diff
@@ -330,6 +331,68 @@ private PdbSymbolReader OpenAssociatedSymbolFile(string peFilePath, PEReader peR
+                    typeOfEmbeddedField = field.FieldType;
+                }
+
+                if ((typeOfEmbeddedField.IsValueType) || (typeOfEmbeddedField.IsPointer))
```
Would we get here with a null `typeOfEmbeddedField` for `[StructLayout(Explicit, Size=4)]struct Foo { }`?

davidwrighton replied:
Sigh, yes, that's a bug.

MichalStrehovsky commented on src/coreclr/src/tools/Common/Compiler/CompilerTypeSystemContext.cs:
```diff
@@ -330,6 +331,68 @@ private PdbSymbolReader OpenAssociatedSymbolFile(string peFilePath, PEReader peR
+                if ((typeOfEmbeddedField.IsValueType) || (typeOfEmbeddedField.IsPointer))
+                {
+                    TypeFlags singleElementFieldType = NormalizedCategoryFor4ByteStructOnX86(typeOfEmbeddedField);
+                    if (singleElementFieldType == TypeFlags.Pointer)
```
Will this do the right thing for `TypeFlags.FunctionPointer`?

davidwrighton replied:
Yes. Single element structures with a function pointer do not receive this optimization

MichalStrehovsky commented on src/coreclr/src/tools/Common/Compiler/CompilerTypeSystemContext.cs:
```diff
@@ -11,6 +11,7 @@
 
 using Internal.TypeSystem;
 using Internal.TypeSystem.Ecma;
+using System.Runtime.InteropServices;
```
Is this needed?

davidwrighton replied:
Nope. Removed

MichalStrehovsky commented on src/coreclr/src/tools/Common/JitInterface/CorInfoImpl.cs:
```diff
@@ -658,6 +658,30 @@ private CorInfoType asCorInfoType(TypeDesc type)
+                        var normalizedCategory = _compilation.TypeSystemContext.NormalizedCategoryFor4ByteStructOnX86(type);
+                        if (normalizedCategory != type.Category)
+                        {
+                            if (NeedsTypeLayoutCheck(type))
```
Can you put this under `#if READYTORUN` to avoid build break in runtimelab?

davidwrighton replied:
Sure. Done.

MichalStrehovsky commented on src/coreclr/src/tools/Common/Compiler/CompilerTypeSystemContext.cs:
```diff
@@ -330,6 +331,68 @@ private PdbSymbolReader OpenAssociatedSymbolFile(string peFilePath, PEReader peR
+                    if (singleElementFieldType == TypeFlags.Pointer)
+                        singleElementFieldType = TypeFlags.UIntPtr;
+                    
+                    switch (singleElementFieldType)
```
Will this do the right thing for enums (`struct MyStruct { MyEnum value } }`)?

davidwrighton replied:
The check above which checked for category being exactly TypeFlags.ValueType was wrong which caused single element structs wrapping an enum to behave incorrectly.

MichalStrehovsky commented on src/coreclr/src/tools/Common/TypeSystem/Common/MetadataFieldLayoutAlgorithm.cs:
```diff
@@ -437,6 +454,7 @@ protected ComputedInstanceFieldLayout ComputeAutoFieldLayout(MetadataType type,
             var layoutMetadata = type.GetClassLayout();
 
             int packingSize = ComputePackingSize(type, layoutMetadata);
+            packingSize = Math.Min(type.Context.Target.MaximumAutoLayoutPackingSize, packingSize);
```
Nit: could you move `TypeSystemContext context = type.Context;` up and use it here?

davidwrighton replied:
Done. I also addressed several other locations within the function which had the same problem

davidwrighton committed:
Fix test oops

davidwrighton committed:
Address code review feedback
Add tests for various single element struct situations
Adjust type system unittests to pass with adjusted X86 type layout rules

davidwrighton committed:
Fix X64 arg iterator handling
